src/services/impl/ThreadPool.cpp

```diff
@@ -0,0 +1,82 @@
+#include "ThreadPool.hh"
+
+#include "RA_Log.h"
+
+namespace ra {
+namespace services {
+namespace impl {
+
+ThreadPool::~ThreadPool()
+{
+    Shutdown(true);
+}
+
+void ThreadPool::Initialize(size_t nThreads) noexcept
+{
+    assert(m_vThreads.empty());
+
+    RA_LOG_INFO("Initializing %zu worker threads", nThreads);
+
+    for (size_t i = 0; i < nThreads; ++i)
+        m_vThreads.emplace_back(&ThreadPool::RunThread, this);
+}
+
+void ThreadPool::RunThread() noexcept
+{
+    do
+    {
+        // check for work
+        do
+        {
+            std::function<void()> pNext;
+            {
+                std::unique_lock<std::mutex> lock(m_oMutex);
+                if (m_vQueue.empty())
+                    break;
+
+                pNext = std::move(m_vQueue.front());
+                m_vQueue.pop();
+            }
+
+            // do work
+            try
+            {
+                pNext();
+            }
+            catch (std::exception ex)
+            {
+                RA_LOG_ERR("Exception on background thread: %s", ex.what());
+            }
+        } while (!m_bShutdownInitiated);
+
+        // wait for work
+        if (!m_bShutdownInitiated)
+        {
+            std::unique_lock<std::mutex> lock(m_oMutex);
+            m_cvMutex.wait(lock);
```

Comment:
The way you have this currently, it might be better to use `std::scoped lock` and change `m_cvMutex` to `std::condition_variable_any` (since C++11, improved in C++14).

Reply:
how is it better? there's plenty of examples of using `unique_lock` and `condition_variable` in this manner, but I can't find any for `scoped_lock` and `condition_variable_any`.

Reply:
Don't worry about it, performance-wise `unique_lock `and `condition_variable `are better. There's not much added benefit unless you need to lock more than one mutex.